pkg/operator/ceph/object/notification/controller.go

```diff
@@ -0,0 +1,227 @@
+		if !cephBN.GetDeletionTimestamp().IsZero() && !cephClusterExists {
+			// Return and do not requeue. Successful deletion.
+			return reconcile.Result{}, nil
+		}
+		logger.Debug("Ceph cluster not yet present.")
+		return reconcileResponse, nil
+	}
+	r.clusterSpec = &cephCluster.Spec
+
+	// DELETE: the CR was deleted
+	if !cephBN.GetDeletionTimestamp().IsZero() {
+		logger.Debugf("CephBucketNotification: %q was deleted", cephBN.Name)
+
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
+	}
+
+	// Populate clusterInfo during each reconcile
+	r.clusterInfo, _, _, err = mon.LoadClusterInfo(r.context, cephCluster.Namespace)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to populate cluster info")
+	}
+
+	// create the notification if needed
+	if err = r.updateCephBucketNotifications(cephBN); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to create bucket notifications")
+	}
+	// Return and do not requeue
+	logger.Debug("CephBucketNotification done reconciling")
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileNotifications) updateCephBucketNotifications(cephBN *cephv1.CephBucketNotification) error {
+	// step 1: fetch all OBCs that has a label matching this CBN
+	namespace := cephBN.Namespace
+	namespaceListOpt := client.InNamespace(namespace)
+	labelListOpt := client.MatchingLabels{
+		notificationLabelPrefix + cephBN.Name: cephBN.Name,
+	}
+	cephOBCList := &bktv1alpha1.ObjectBucketClaimList{}
+	err := r.client.List(context.TODO(), cephOBCList, namespaceListOpt, labelListOpt)
+	if err != nil {
+		return err
+	}
+	if len(cephOBCList.Items) == 0 {
+		logger.Debugf("no ObjectbucketClaim associated with notification: %q", cephBN.Name)
+		return nil
+	}
+
+	// step 2: get the topic associated with the notification
+	bucketTopic := &cephv1.CephBucketTopic{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephBN.Spec.Topic}, bucketTopic); err != nil {
+		return errors.Wrapf(err, "failed to retrieve CephBucketTopic %q", cephBN.Spec.Topic)
+	}
+	topicARN, err := topic.GetARN(bucketTopic)
+	if err != nil {
+		return errors.Wrap(err, "failed to get topic ARN")
+	}
+
+	// step 3: loop through all OBCs in the list and get their OB and SC
+	for _, cephOBC := range cephOBCList.Items {
+		cephOB := bktv1alpha1.ObjectBucket{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.ObjectBucketName}, &cephOB); err != nil {
+			return errors.Wrapf(err, "failed to retrieve ObjectBucket %q", cephOBC.Spec.ObjectBucketName)
+		}
+		cephSC := storagev1.StorageClass{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.StorageClassName}, &cephSC); err != nil {
+			return errors.Wrapf(err, "failed to retrieve StorageClass %q", cephOBC.Spec.StorageClassName)
+		}
```

Comment:
A storage class is only used when an OBC creates an OB. After that, I think we can no longer consider the SC to be a source of truth. It might be modified or even deleted. We should probably leave out any investigation of the SC.

Reply:
then how do we verify that the bucket and notification belong to the same object store?

Reply:
Because we cannot rely on the StorageClass as a source of truth, we have to either...
1. have the user provide the object store name and namespace in the CBN resource, OR
2. inspect the OB and/or OBC to get the information